presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java

```diff
@@ -0,0 +1,282 @@
+    private final ScheduledExecutorService updateScheduledExecutor;
+    private final JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec;
+    private final JsonCodec<PlanFragment> planFragmentCodec;
+    private final JsonCodec<TaskInfo> taskInfoCodec;
+
+    @GuardedBy("this")
+    private TaskInfo taskInfo;
+    @GuardedBy("this")
+    private ScheduledFuture<?> scheduledFuture;
+
+    public NativeExecutionTask(
+            Session session,
+            TaskId taskId,
+            PlanFragment planFragment,
+            List<TaskSource> sources,
+            HttpClient httpClient,
+            TableWriteInfo tableWriteInfo,
+            RemoteTaskStats stats,
+            Executor executor,
+            ScheduledExecutorService updateScheduledExecutor,
+            JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec,
+            JsonCodec<PlanFragment> planFragmentCodec,
+            JsonCodec<TaskInfo> taskInfoCodec)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(taskId, "taskId is null");
+        requireNonNull(planFragment, "planFragment is null");
+        requireNonNull(sources, "sources is null");
+        requireNonNull(httpClient, "httpClient is null");
+        requireNonNull(tableWriteInfo, "tableWriteInfo is null");
+        requireNonNull(stats, "stats is null");
+        requireNonNull(executor, "executor is null");
+        requireNonNull(updateScheduledExecutor, "updateScheduledExecutor is null");
+        requireNonNull(taskUpdateRequestCodec, "taskUpdateRequestCodec is null");
+        requireNonNull(planFragmentCodec, "planFragmentCodec is null");
+        requireNonNull(taskInfoCodec, "taskInfoCodec is null");
+
+        try (SetThreadName ignored = new SetThreadName("NativeExecutionTask-%s", taskId)) {
+            this.taskId = taskId;
+            this.session = session;
+            this.planFragment = planFragment;
+            this.httpClient = httpClient;
+            this.tableWriteInfo = tableWriteInfo;
+            this.stats = stats;
+            this.sources = sources;
+            this.executor = executor;
+            this.updateScheduledExecutor = updateScheduledExecutor;
+            this.outputBuffers = createInitialEmptyOutputBuffers(PARTITIONED);
+            List<BufferInfo> bufferStates = outputBuffers.getBuffers()
+                    .keySet().stream()
+                    .map(outputId -> new BufferInfo(outputId, false, 0, 0, PageBufferInfo.empty()))
+                    .collect(toImmutableList());
+            this.taskInfo = createInitialTask(taskId, URI.create(LOCAL_URI + taskId), bufferStates, new TaskStats(DateTime.now(), null), "");
+            this.taskUpdateRequestCodec = taskUpdateRequestCodec;
+            this.planFragmentCodec = planFragmentCodec;
+            this.taskInfoCodec = taskInfoCodec;
+        }
+    }
+
+    public TaskInfo getTaskInfo()
+    {
+        synchronized (taskInfo) {
+            return taskInfo;
+        }
```

Comment:
After returning taskInfo is not protected anymore at callsite. You have to make a deep copy in the synchronized block and return the copy

Reply:
TaskInfo itself is an immutable object, so there is no need to deep copy.  I'm wondering what kind of concurrency requirements we have for the TaskInfo object and the read access requests.  How important is it for reads and writes to be synchronized?  Looks like there's only one method that sets taskInfo and it's a simple set operation.  Maybe making the field volatile would do the trick.

Reply:
Ah.. Then it makes sense.